Sources/Structures/DataLoading/PaginationDataLoading/PaginationWrapperDelegate.swift

```diff
@@ -0,0 +1,96 @@
+//
+//  THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+/// PaginationWrapper delegate used for pagination results handling and
+/// customization of bound states (loading, empty, error, etc.).
+public protocol PaginationWrapperDelegate: class {
+
+    associatedtype DataSourceType: DataSource
+
+    /// Delegate method that handles loading new chunk of data.
+    ///
+    /// - Parameters:
+    ///   - newItems: New items.
+    ///   - dataSource: Data source used to load items
+    func paginationWrapper(didLoad newItems: DataSourceType.ResultType,
+                           using dataSource: DataSourceType)
+
+    /// Delegate method that handles reloading or initial loading of data.
+    ///
+    /// - Parameters:
+    ///   - allItems: New items.
+    ///   - dataSource: Data source used to load items
+    func paginationWrapper(didReload allItems: DataSourceType.ResultType,
+                           using dataSource: DataSourceType)
+
+    /// Delegate method that returns placeholder view for empty state.
+    ///
+    /// - Returns: Configured instace of UIView.
+    func emptyPlaceholder() -> UIView
+
+    /// Delegate method that is called when initial loading error is occured.
+    /// It should return true if error is handled and false if it doesn't.
+    ///
+    /// - Parameters:
+    ///   - error: Error that occured due data loading.
+    /// - Returns: Bool value. If true - then error placeholder wouldn't be shown.
+    func customInitialLoadingErrorHandling(for error: Error) -> Bool
+
+    /// Delegate method that returns placeholder view for error state.
+    ///
+    /// - Parameters:
+    ///   - error: Error that occured due data loading.
+    /// - Returns: Configured instace of UIView.
+    func errorPlaceholder(for error: Error) -> UIView
+
+    /// Delegate method that returns loading idicator for initial loading state.
+    /// This indicator will appear at center of the placeholders container.
+    ///
+    /// - Returns: Configured instace of AnyLoadingIndicator.
+    func initialLoadingIndicator() -> AnyLoadingIndicator
+
+    /// Delegate method that returns loading idicator for initial loading state.
+    ///
+    /// - Returns: Configured instace of AnyLoadingIndicator.
+    func loadingMoreIndicator() -> AnyLoadingIndicator
+
+    /// Delegate method that returns instance of UIButton for "retry load more" action.
+    ///
+    /// - Returns: Configured instace of AnyLoadingIndicator.
+    func retryLoadMoreButton() -> UIButton
```

Comment:
Bad naming. `Retry button` or `Load more button`. Wrong description. Fix it, please